pinot-plugins/pinot-input-format/pinot-csv/src/main/java/org/apache/pinot/plugin/inputformat/csv/CSVDefaultHeaderUtil.java

```diff
@@ -0,0 +1,195 @@
+        _dataType = Type.ARRAY;
+        return;
+      } else if (NumberUtils.isParsable(value.toString())) {
+        _dataType = Type.NUMBER;
+      }
+      _length = value.toString().length();
+    }
+
+    public Type getDataType() {
+      return _dataType;
+    }
+
+    public int getLength() {
+      return _length;
+    }
+  }
+
+  /**
+   * Detects the types for each column based on the row values. If any column is of a single type (e.g. number), except
+   * for the first row, then the first row is presumed to be a header. If the type of the column is assumed to
+   * be a string, the length of the strings within the body is the determining factor. Specifically, if all the rows
+   * except the first are the same length, it's a header.
+   *
+   * If no header is detected, then the header property within the CSVRecordReaderConfig will be set with the following
+   * e.g. {col_0, col_1, col_2, ..}. If a header is detected within the file, then the header property will not
+   * be set and will default to use the first row of the file.
+   *
+   * Motivation for this logic is taken from: https://github.com/python/cpython/blob/main/Lib/csv.py
+   *
+   * Returns the default header when no header is detected. If the header is detected, then returns null.
+   */
+  public static String getDefaultHeader(File csvFile, CSVFormat format, Character delimiter,
+      Character multiValueDelimiter)
+      throws IOException {
+    // Assume there is no header and the default header based on number of columns. We need the default header or
+    // else the CSVRecordReader may not read the values correctly when there are columns with the same values.
+    int numColumns = 0;
+    int rowCounter = 0;
+    List<CSVRecord> records = new ArrayList<>();
+    String[] originalHeader = format.getHeader();
+    try (BufferedReader bufferedReader = RecordReaderUtils.getBufferedReader(csvFile)) {
+      format.withHeader(null);
+      CSVParser parser = format.parse(bufferedReader);
+      Iterator<CSVRecord> iterator = parser.iterator();
+      while (iterator.hasNext() && rowCounter < MAX_CSV_SAMPLE_ROWS + 1) {
+        CSVRecord row = iterator.next();
+        records.add(row);
+        int curNumColumns = row.size();
+        if (curNumColumns > numColumns) {
+          numColumns = curNumColumns;
+        }
+        rowCounter++;
+      }
+    }
+
+    if (records.size() == 0) {
+      throw new IllegalStateException("CSV file is empty. csvFile=" + csvFile.getAbsolutePath());
+    }
+
+    String[] defaultColumnNames = new String[numColumns];
+    for (int i = 0; i < numColumns; i++) {
+      defaultColumnNames[i] = DEFAULT_CSV_COLUMN_PREFIX + i;
+    }
+
+    HashMap<String, ColumnValueAttributes> csvHeaderAttributes = new HashMap<>(numColumns);
+    HashMap<String, ColumnValueAttributes> csvBodyAttributes = new HashMap<>(numColumns);
+    Iterator<CSVRecord> iterator = records.iterator();
+    if (iterator.hasNext()) {
```

Comment:
I assume this check is not needed since we already have size check previously.

Reply:
good catch. updated.